### PR TITLE
Unit test for consents backend functions

### PR DESCRIPTION
**Pull request by xmano**
Add unit test in backend-function apps

### CODE REVIEW

**Comment by xmano** [comment]
Need to isolate `db.runTransaction()` if it is causing issue..

File: apps/backend-functions/src/consent.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export const createConsent = async (
 
     const consentSnap = await tx.get(db.doc(`consents/${consent.id}`));
     const consentData = consentSnap.data();
-
+
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.ts:51]: remove extra space/tabs (use formatter)

And this file have no changes

File: apps/backend-functions/src/fixtures/inviteUsers.ts [hunk 2]
```diff
@@ -19,8 +19,15 @@ export const testFixture = {
...
+    lastName: 'Two',
     email: 'u2@cascade8.com',
     uid: 'uid-user2',
     orgId: 'O001',
   },
+
+  //Consents
+  'consents/O001': {
+    id: 'O001',
+  }
```
Comment (phpgeekfr) [inline apps/backend-functions/src/fixtures/inviteUsers.ts:32]: please create a distinct file for consents, the file is named `inviteUsers`

File: apps/backend-functions/src/consent.spec.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+import * as userOps from './internals/users';
+import { firebase } from '@env';
+import { expect } from '@jest/globals';
+import { ErrorResultResponse } from './utils';
+import { ConsentType } from '@blockframes/consents/+state/consents.firestore';
+import { endMaintenance } from '@blockframes/firebase-utils';
+
+const testEnv = firebaseTest(firebase());
+
+describe('Invitation backend-function unit-tests', () => {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:17]: bad naming

File: apps/backend-functions/src/consent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+import { createConsent } from './main';
+import firebaseTest = require('firebase-functions-test');
+import { testFixture } from './fixtures/inviteUsers';
+import * as admin from 'firebase-admin';
+import * as userOps from './internals/users';
+import { firebase } from '@env';
+import { expect } from '@jest/globals';
+import { ErrorResultResponse } from './utils';
+import { ConsentType } from '@blockframes/consents/+state/consents.firestore';
+import { endMaintenance } from '@blockframes/firebase-utils';
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:13]: unused imports: 
```
import * as userOps from './internals/users';
import { ErrorResultResponse } from './utils';
import { ConsentType } from '@blockframes/consents/+state/consents.firestore';
```

File: apps/backend-functions/src/consent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    // After all tests, db is reseted
+    await clearFirestoreData({ projectId: firebase().projectId });
+  });
+
+  describe('Consents spec', () => {
+    it('missing auth context, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:34]: in `apps/backend-functions/src/consent.ts` create an interface :

```
export interface ConsentData  { consentType: ConsentType; ip: string; docId: string; filePath?: string } ;
```

use it in 
```
 */
export const createConsent = async (
  data: ConsentData,
  context: CallableContext
)
```

and use it here to type `data`:

`const data : ConsentData  = { ...}`

File: apps/backend-functions/src/consent.spec.ts [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, {}))
+            .rejects
+            .toThrow('Permission denied: missing auth context.');
+    });
+
+    it('missing user data, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:50]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, context))
+            .rejects
+            .toThrow('Invalid user');
+    });
+
+    it('missing org ID, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:73]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, context))
+            .rejects
+            .toThrow('Invalid organization');
+    });
+
+    it('missing docID in data param, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:96]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, context))
+            .rejects
+            .toThrow('Undefined docId');
+    });
+
+    it('missing ip in data param, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:119]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, context))
+            .rejects
+            .toThrow('Undefined ip');
+    });
+
+    it('missing filePath in data param, throws error', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:142]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+      await expect(wrapped(data, context))
+            .rejects
+            .toThrow('Invalid filePath');
+    });
+
+    it('creates \'Access Type\' consents document', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:165]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+          ip: "10.0.0.1"
+        })
+      );
+    });
+
+    it('creates \'Share Type\' consents document', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:207]: const data : ConsentData = { ...}

File: apps/backend-functions/src/consent.spec.ts [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+      const result = await wrapped(data, context);
+      expect(result).toBeTruthy();
+
+      //Fetch the data and compare the result
+      const snap = await admin.firestore().collection('consents').doc(data.docId).get();
+      const consentsData = snap.data();
+      expect(consentsData.access[0]).toEqual(
+        expect.objectContaining({
+          docId: "O001",
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:229]: use single quotes

File: apps/backend-functions/src/consent.spec.ts [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+          lastName: "Two",
+          ip: "10.0.0.1"
+        })
+      );
+
+      expect(consentsData.share[0]).toEqual(
+        expect.objectContaining({
+          docId: "O001",
+          email: "u2@cascade8.com",
+          firstName: "User",
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:241]: use single quotes

File: apps/backend-functions/src/consent.spec.ts [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+      );
+    });
+
+    it('creates \'Share Type\' consents document', async () => {
+      const wrapped = testEnv.wrap(createConsent);
+
+      //Compose the call to simpleCallable cf with param data
+      const data = {
+        consentType: 'share',
+        filePath: '/c/o/marketplace',
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:209]: `filePath` should not be needed for `share` type

File: apps/backend-functions/src/consent.spec.ts [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        expect.objectContaining({
+          id:"O001",
+          share: []
+        })
+      );
+
+      expect(consentsData.access[0]).toEqual(
+        expect.objectContaining({
+          docId: "O001",
+          email: "u2@cascade8.com",
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:195]: use single quotes

File: apps/backend-functions/src/fixtures/inviteUsers.ts (hunk 2 above)
Comment (phpgeekfr) [reply apps/backend-functions/src/fixtures/inviteUsers.ts:32]: or rename it

File: apps/backend-functions/src/consent.spec.ts (hunk 14 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:241]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 16 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:195]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 13 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:229]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 4 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:13]: removed

File: apps/backend-functions/src/consent.spec.ts (hunk 3 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:17]: fixed

File: apps/backend-functions/src/consent.spec.ts (hunk 5 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:34]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 6 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:50]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 7 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:73]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 8 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:96]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 9 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:119]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 10 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:142]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 11 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:165]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 12 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:207]: done

File: apps/backend-functions/src/consent.spec.ts (hunk 15 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts:209]: ok.. removed this.

File: apps/backend-functions/src/consent.ts (hunk 1 above)
Comment (xmano) [reply apps/backend-functions/src/consent.ts:51]: done

File: apps/backend-functions/src/fixtures/inviteUsers.ts (hunk 2 above)
Comment (xmano) [reply apps/backend-functions/src/fixtures/inviteUsers.ts:32]: As discussed i will be using a common file data.ts here

File: apps/backend-functions/src/consent.spec.ts [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+      const consentsData = snap.data();
+      expect(consentsData.access[0]).toEqual(
+        expect.objectContaining({
+          docId: 'O001',
+          email: 'u2@cascade8.com',
+          firstName: 'User',
+          lastName: 'Two',
+          ip: '10.0.0.1'
+        })
+      );
```
Comment (phpgeekfr) [inline apps/backend-functions/src/consent.spec.ts:232]: This data is coming from previous test. so If I do `it.only` on this one it fail

**Review by phpgeekfr — APPROVED** [review]

File: apps/backend-functions/src/consent.spec.ts (hunk 17 above)
Comment (xmano) [reply apps/backend-functions/src/consent.spec.ts]: fixed

**Pull request merged**